### PR TITLE
[feat](mtmv)Paimon queries the data in the cache instead of querying the latest data

### PR DESCRIPTION
**Pull request by zddr**
### What problem does this PR solve?

- add `PaimonMetadataCacheMgr` in `ExternalMetaCacheMgr` to manage snapshotCache of paimon table
- move paimonSchemaCache to PaimonMetadataCacheMgr, and add schemaId as part of key
- PaimonExternalTable overrides the methods in ExternalTable and supports partition pruning
- PaimonExternalTable implements the MvcTable interface, supporting the retrieval of snapshot data from the cache during queries to avoid cache refreshes that may result in different versions of metadata being used in a single query
- MTMVTask retrieves snapshot data of mvccTable before the task starts to avoid cache refresh that may result in different versions of metadata being used in a single refresh task

**behavior changes of query  paimon table:**
- FE has just started and is query the latest data
- Paimon data has changed, Doris is still query the previous data
- After the snapshot cache expires, Doris will query the latest data
- desc paimon; The schema corresponding to the snapshotId in the snapshot cache is displayed

Issue Number: close #xxx

Related PR: #xxx

Problem Summary:
Paimon queries the data in the cache instead of querying the latest data
### Release note
Paimon queries the data in the cache instead of querying the latest data

### Check List (For Author)

- Test 
    - [x] Regression test
    - [ ] Unit Test
    - [ ] Manual test (add detailed scripts or steps below)
    - [ ] No need to test or manual test. Explain why:
        - [ ] This is a refactor/code format and no logic has been changed.
        - [ ] Previous test can cover this change.
        - [ ] No code files have been changed.
        - [x] Other reason 
        The previous case was able to cover normal read and write operations, but querying fixed versions of cases requires the ability to manipulate Paimon data, which is currently not possible

- Behavior changed:
    - [ ] No.
    - [x] Yes. 

- Does this need documentation?
    - [ ] No.
    - [x] Yes. 

### Check List (For Reviewer who merge this PR)

- [ ] Confirm the release note
- [ ] Confirm test cases
- [ ] Confirm document
- [ ] Add branch pick label

### CODE REVIEW

**Comment by zddr** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/datasource/paimon/PaimonMetadataCacheMgr.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        return paimonMetadataCache;
+    }
+
+    public void removeCache(long catalogId) {
+        paimonMetadataCache.invalidateCatalogCache(catalogId);
+    }
+
+    public void invalidateCatalogCache(long catalogId) {
+        paimonMetadataCache.invalidateCatalogCache(catalogId);
+    }
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/datasource/paimon/PaimonMetadataCacheMgr.java:40]: what's different between these two interface

File: fe/fe-core/src/main/java/org/apache/doris/datasource/paimon/source/PaimonSource.java [hunk 2]
```diff
@@ -36,7 +37,8 @@ public class PaimonSource {
     public PaimonSource(TupleDescriptor desc) {
         this.desc = desc;
         this.paimonExtTable = (PaimonExternalTable) desc.getTable();
-        this.originTable = paimonExtTable.getPaimonTable();
+        this.originTable = paimonExtTable.getPaimonTable(
+                ConnectContext.get().getStatementContext().getSnapshot(paimonExtTable));
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/datasource/paimon/source/PaimonSource.java:41]: do we need check `ConnectContext.get() == null` and `ConnectContext.get().getStatementContext()`. sometimes they do not exist

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PruneFileScanPartition.java [hunk 3]
```diff
@@ -74,17 +74,17 @@ public Rule build() {
     private SelectedPartitions pruneExternalPartitions(ExternalTable externalTable,
             LogicalFilter<LogicalFileScan> filter, LogicalFileScan scan, CascadesContext ctx) {
         Map<String, PartitionItem> selectedPartitionItems = Maps.newHashMap();
-        // todo: real snapshotId
-        if (CollectionUtils.isEmpty(externalTable.getPartitionColumns(Optional.empty()))) {
+        if (CollectionUtils.isEmpty(externalTable.getPartitionColumns(
+                ConnectContext.get().getStatementContext().getSnapshot(externalTable)))) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PruneFileScanPartition.java:78]: get statementContext from cascadescontext

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PruneFileScanPartition.java [hunk 4]
```diff
@@ -74,17 +74,17 @@ public Rule build() {
...
             // non partition table will be handled in HiveScanNode.
             return SelectedPartitions.NOT_PRUNED;
         }
         Map<String, Slot> scanOutput = scan.getOutput()
                 .stream()
                 .collect(Collectors.toMap(slot -> slot.getName().toLowerCase(), Function.identity()));
-        // todo: real snapshotId
-        List<Slot> partitionSlots = externalTable.getPartitionColumns(Optional.empty())
+        List<Slot> partitionSlots = externalTable.getPartitionColumns(
+                        ConnectContext.get().getStatementContext().getSnapshot(externalTable))
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PruneFileScanPartition.java:87]: ditto

File: fe/fe-core/src/main/java/org/apache/doris/nereids/StatementContext.java [hunk 5]
```diff
@@ -526,19 +526,37 @@ public void loadSnapshots(Map<List<String>, TableIf> tables) {
...
      * @return MvccSnapshot
      */
-    public MvccSnapshot getSnapshot(MvccTable mvccTable) {
-        return snapshots.get(new MvccTableInfo(mvccTable));
+    public Optional<MvccSnapshot> getSnapshot(TableIf tableIf) {
+        if (!(tableIf instanceof MvccTable)) {
+            return Optional.empty();
+        }
+        MvccTableInfo mvccTableInfo = new MvccTableInfo(tableIf);
+        return snapshots.containsKey(mvccTableInfo) ? Optional.of(snapshots.get(mvccTableInfo)) : Optional.empty();
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/StatementContext.java:549]: could simplify by `Optional.ofNullable(snapshots.get(mvccTableInfo))`

File: fe/fe-core/src/main/java/org/apache/doris/datasource/paimon/PaimonUtil.java [hunk 6]
```diff
@@ -152,4 +169,107 @@ public static ListPartitionItem toListPartitionItem(String partitionName, List<C
...
+                        tsScale = 6;
+                    }
+                } else if (dataType instanceof org.apache.paimon.types.LocalZonedTimestampType) {
+                    tsScale = ((org.apache.paimon.types.LocalZonedTimestampType) dataType).getPrecision();
+                    if (tsScale > 6) {
+                        tsScale = 6;
+                    }
+                }
+                return ScalarType.createDatetimeV2Type(tsScale);
+            case TIMESTAMP_WITH_LOCAL_TIME_ZONE:
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/datasource/paimon/PaimonUtil.java:213]: so data load into doris will lead to wrong time zone?

Comment (zddr) [reply fe/fe-core/src/main/java/org/apache/doris/datasource/paimon/PaimonUtil.java:213]: This is not a newly added method, it was copied from paimonExternalTable

File: fe/fe-core/src/main/java/org/apache/doris/datasource/paimon/PaimonMetadataCacheMgr.java (hunk 1 above)
Comment (zddr) [reply fe/fe-core/src/main/java/org/apache/doris/datasource/paimon/PaimonMetadataCacheMgr.java:40]: Just the caller is different, one is called when deleting the catalog, and the other is called when refreshing the catalog. Other MGRs provide two methods, so Paimon also provides two methods with the same logic here

**Comment by zddr** [comment]
run buildall

**Review by morrySnow — DISMISSED** [review]

**Review by Jibing-Li — DISMISSED** [review]

File: fe/fe-core/src/main/java/org/apache/doris/datasource/paimon/PaimonMetadataCache.java [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+        }
+        throw new CacheException("failed to loadSchema for: %s.%s.%s.%s",
+                null, key.getCatalog().getName(), key.getDbName(), key.getTableName(), key.getSchemaId());
+    }
+
+    @NotNull
+    private PaimonSnapshotCacheValue loadSnapshot(PaimonSnapshotCacheKey key) {
+        try {
+            PaimonSnapshot latestSnapshot = loadLatestSnapshot(key);
+            List<Column> partitionColumns = getPaimonSchema(key.getCatalog(), key.getDbName(), key.getTableName(),
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/datasource/paimon/PaimonMetadataCache.java:119]: I think this is not safe to get value from another cache when loading cache.
Both cache use some executor pool and may becausing dead lock

File: fe/fe-core/src/main/java/org/apache/doris/datasource/paimon/PaimonMetadataCache.java [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.OptionalLong;
+import java.util.Set;
+import java.util.concurrent.ExecutorService;
+
+public class PaimonMetadataCache {
+
+    private final LoadingCache<PaimonSnapshotCacheKey, PaimonSnapshotCacheValue> snapshotCache;
+    private final LoadingCache<PaimonSchemaCacheKey, PaimonSchemaCacheValue> schemaCache;
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/datasource/paimon/PaimonMetadataCache.java:54]: What is the relationship between this and ExternalSchemaCache?
And why not put this schemCache in ExternalSchemaCache?

**Review by morningman — CHANGES_REQUESTED** [review]

**Review by wuwenchi — APPROVED** [review]

**Comment by zddr** [comment]
run buildall

**Comment by zddr** [comment]
run buildall

**Comment by zddr** [comment]
run p0

**Review by morningman — APPROVED** [review]

**Pull request merged**